newIDE/app/src/ProjectCreation/NewProjectSetupDialog.js

```diff
@@ -0,0 +1,241 @@
+  storageProviders: Array<StorageProvider>,
+  authenticatedUser: AuthenticatedUser,
+  onChangeSubscription: () => void,
+|};
+
+const generateProjectName = (sourceExampleName: ?string) =>
+  sourceExampleName
+    ? `${generateName()} (${sourceExampleName})`
+    : generateName();
+
+const NewProjectSetupDialog = ({
+  isOpening,
+  onClose,
+  onCreate,
+  sourceExampleName,
+  storageProviders,
+  authenticatedUser,
+  onChangeSubscription,
+}: Props): React.Node => {
+  const [projectNameError, setProjectNameError] = React.useState<?React.Node>(
+    null
+  );
+  const [projectName, setProjectName] = React.useState<string>(() =>
+    generateProjectName(sourceExampleName)
+  );
+  const [saveAsLocation, setSaveAsLocation] = React.useState<?SaveAsLocation>(
+    null
+  );
+  const [storageProvider, setStorageProvider] = React.useState<StorageProvider>(
+    () => {
+      if (authenticatedUser.authenticated) {
+        const cloudStorageProvider = storageProviders.find(
+          ({ internalName }) => internalName === 'Cloud'
+        );
+        if (cloudStorageProvider) return cloudStorageProvider;
+      }
+
+      const localFileStorageProvider = storageProviders.find(
+        ({ internalName }) => internalName === 'LocalFile'
+      );
+      if (localFileStorageProvider) return localFileStorageProvider;
+
+      return emptyStorageProvider;
+    }
+  );
+
+  const needUserAuthentication =
+    storageProvider.needUserAuthentication && !authenticatedUser.authenticated;
+  const { limits } = authenticatedUser;
+  const hasTooManyCloudProjects =
+    storageProvider.internalName === 'Cloud' &&
+    checkIfHasTooManyCloudProjects(authenticatedUser);
+
+  const onValidate = React.useCallback(
+    async () => {
+      if (isOpening) return;
+      if (needUserAuthentication) return;
+
+      setProjectNameError(null);
+      if (!projectName) {
+        setProjectNameError(
+          <Trans>Please enter a name for your project.</Trans>
+        );
+        return;
+      }
+      await onCreate({ projectName, storageProvider, saveAsLocation });
+    },
+    [
+      isOpening,
+      onCreate,
+      needUserAuthentication,
+      projectName,
+      storageProvider,
+      saveAsLocation,
+    ]
+  );
+
+  const _onChangeProjectName = React.useCallback(
+    (event, text) => {
+      if (projectNameError) setProjectNameError(null);
+      setProjectName(text);
+    },
+    [setProjectName, projectNameError]
+  );
+
+  return (
+    <Dialog
+      open
+      title={<Trans>New Project</Trans>}
+      id="project-pre-creation-dialog"
+      maxWidth="sm"
+      actions={[
+        <FlatButton
+          disabled={isOpening}
+          key="cancel"
+          label={<Trans>Cancel</Trans>}
+          onClick={onClose}
+        />,
+        <LeftLoader isLoading={isOpening}>
+          <DialogPrimaryButton
+            primary
+            disabled={
+              isOpening || needUserAuthentication || hasTooManyCloudProjects
```

Comment:
can't create a project if too many cloud? 
Even if it will not be saved afterwards?

Reply:
`hasTooManyCloudProjects` will be true only if the storageProvider is Cloud - so if you have too many cloud projects you can just choose "Don't save for now" and do some cleanup/save/download as a file later :)